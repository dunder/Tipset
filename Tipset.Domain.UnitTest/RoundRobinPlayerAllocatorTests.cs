using System;
using System.Collections.Generic;
using ApprovalTests;
using ApprovalTests.Reporters;
using NUnit.Framework;

namespace Tipset.Domain.UnitTest
{

    [TestFixture]
    public class RoundRobinPlayerAllocatorTests
    {
        private RoundRobinPlayerAllocator allocator;

        [TestFixtureSetUp]
        public void SetupFixture()
        {
            allocator = new RoundRobinPlayerAllocator();
        }

        [Test]
        public void PlayersArgumentNullThrows()
        {
            var exception = Assert.Throws<ArgumentException>(() => allocator.Allocate(null, new List<Round>()));
            Assert.That(exception.Message, Is.EqualTo("players"));
        }

        [Test]
        public void RoundsArgumentNullShouldThrow()
        {
            var exception = Assert.Throws<ArgumentException>(() => allocator.Allocate(new List<Player>(), null));
            Assert.That(exception.Message, Is.EqualTo("rounds"));
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void ShouldAllocatePlayersInARoundRobinFashion()
        {
            var players = new List<Player>
                      {
                          new Player {Id = "1"},
                          new Player {Id = "2"},
                          new Player {Id = "3"}
                      };

            IList<Round> rounds = new List<Round>
                                  {
                                      new Round {Date = new DateTime(2013,5,1)},
                                      new Round {Date = new DateTime(2013,5,2)},
                                      new Round {Date = new DateTime(2013,5,3)},
                                      new Round {Date = new DateTime(2013,5,4)},
                                      new Round {Date = new DateTime(2013,5,5)},
                                      new Round {Date = new DateTime(2013,5,6)},
                                  };

            rounds = allocator.Allocate(players, rounds);

            Approvals.VerifyAll(rounds, "round");
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void ShouldAllocatePlayersInARoundRobinFashionAndAssignSurplusToLastRound()
        {
            var players = new List<Player>
                      {
                          new Player {Id = "1"},
                          new Player {Id = "2"},
                          new Player {Id = "3"}
                      };

            IList<Round> rounds = new List<Round>
                                  {
                                      new Round {Date = new DateTime(2013,5,1)},
                                      new Round {Date = new DateTime(2013,5,2)},
                                      new Round {Date = new DateTime(2013,5,3)},
                                      new Round {Date = new DateTime(2013,5,4)},
                                      new Round {Date = new DateTime(2013,5,5)},
                                  };

            rounds = allocator.Allocate(players, rounds);

            Approvals.VerifyAll(rounds, "round");
        }
    }
}
