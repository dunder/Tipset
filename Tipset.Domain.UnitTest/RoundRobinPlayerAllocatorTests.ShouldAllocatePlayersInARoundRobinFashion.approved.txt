round[0] = Date: 2013-05-01, Assigned player ID:s: 1
round[1] = Date: 2013-05-02, Assigned player ID:s: 2
round[2] = Date: 2013-05-03, Assigned player ID:s: 3
round[3] = Date: 2013-05-04, Assigned player ID:s: 1
round[4] = Date: 2013-05-05, Assigned player ID:s: 2
round[5] = Date: 2013-05-06, Assigned player ID:s: 3
